### PR TITLE
Add ability to easily make singleton classes

### PR DESCRIPTION
**Pull request by gappleto97**
This PR allows users to quickly generate singleton classes. This is useful when objects are getting serialized or deserialized, and can yield a large performance benefit.

Each singleton is associated with the class constructor and the list of attributes used. So while `Example(3) is Example(3)` will always be `True`, `Example(3) is Example(4)` will always be `False`,
as would `Example(3) is InheritsFromExample(3)`.

This functionality is implemented by adding a `__new__()` method to the generated class, as well as a dictionary `_singletons`. Each instance also gets a property `_skip_init` to short circuit the `__init__` method if called again on the same instance (because `__init__` is called on *every* instance returned from `__new__`).

~~I have not yet been able to make this work in Python 2.7. It yields the error: `TypeError: unbound method __new__() must be called with C instance as first argument (got type instance instead)`~~

In addition, I am not entirely sure what format you want the changelog items to be in.

# Pull Request Check List

This is just a reminder about the most common mistakes.  Please make sure that you tick all *appropriate* boxes.  But please read our [contribution guide](https://www.attrs.org/en/latest/contributing.html) at least once, it will save you unnecessary review cycles!

- [x] Added **tests** for changed code.
- [ ] New features have been added to our [Hypothesis testing strategy](https://github.com/python-attrs/attrs/blob/master/tests/strategies.py).
- [x] Changes or additions to public APIs are reflected in our type stubs (files ending in ``.pyi``).
    - [x] ...and used in the stub test file `tests/typing_example.py`.
- [x] Updated **documentation** for changed code.
    - [x] New functions/classes have to be added to `docs/api.rst` by hand.
    - [x] Changed/added classes/methods/functions have appropriate `versionadded`, `versionchanged`, or `deprecated` [directives](http://www.sphinx-doc.org/en/stable/markup/para.html#directive-versionadded).
- [x] Documentation in `.rst` files is written using [semantic newlines](http://rhodesmill.org/brandon/2012/one-sentence-per-line/).
- [ ] Changes (and possible deprecations) have news fragments in [`changelog.d`](https://github.com/python-attrs/attrs/blob/master/changelog.d).

If you have *any* questions to *any* of the points above, just **submit and ask**!  This checklist is here to *help* you, not to deter you from contributing!

### CODE REVIEW

File: tests/test_make.py [hunk 1]
```diff
@@ -632,6 +632,23 @@ def test_sugar_callable(self):
...
 
+    def test_singleton(self):
+        """
+        Ensure that classes can be made singletons
+        """
+        @attr.s(singleton=True)
+        class C(object):
+            x = attr.ib(default=3)
+
+        a = C()
```
[the remark this answers is not in the document]
Comment (gappleto97) [reply tests/test_make.py]: This has been addressed in my latest commit. I think there is a way to make it implied if `frozen` is not specified, but I did not want to make it start as `None` if that might break other things. If I can get confirmation that this is okay, I will change it to be that way.

**Comment by gappleto97** [comment]
On lunch, so I figured I would take another look at the python2 error. It looks like the `TypeError`  is coming from the fact that you can't successfully assign to `__new__` on a python2 class. Possible solutions:

1. Add a special case to build a new stub class and inherit from the one given to the decorator
2. Add a special case similar to `__slots__` where it generates a new class entirely
3. Say this feature is python3 only to avoid special cases

I will await further guidance on which to implement.

**Comment by gappleto97** [comment]
Just added a fix for python2. Turns out that python3 was implicitly making it a staticmethod, as if it were declared in the class itself. Python2 was not doing this, so it has to be explicitly declared as a staticmethod

**Comment by gappleto97** [comment]
@RonnyPfannschmidt I believe this is now ready for a final review

**Comment by hynek** [comment]
This honestly looks a bit random/wrong level of abstraction. I could live with adding it to attrs but I really think it should be a separate decorator.

Do you think you could do that?

**Comment by gappleto97** [comment]
@hynek I do think that I could do that, but I'm curious as to why you think it's at the wrong level of abstraction? The goal I have in mind here is to make unique pieces of data be consistently the same object. This not only improves performance, but it's also hard to do in other places of the library.

Particularly, this idea came from code that I have been working on at work. I keep having to deal with serialized identifiers that are appearing several times in the same file, and this means that I have a fair bit of boilerplate that I need to maintain. The version implemented in this changeset is more efficient than that because it can work with the internals of attrs to make it happen. I know that I have had this as a use case in the past as well.

As for why it's more difficult in other places, the easiest example is adding the `__init__` short circuit. There are definitely solutions, but they all involve less direct methods than was used in this patch set.

Basically, I can't think of any case where I would really want to use the singleton portion and *not* the rest of it as well. I'm curious as to where you disagree.

**Comment by Tinche** [comment]
Honestly this seems interesting but it's kind of a niche thing. Speaking from personal experience, I've literally never needed this functionality. I know the plural of anecdote is not data, but my gut feeling is this is too niche to have on `attr.s()`. Hence the idea about a different decorator.

**Comment by gappleto97** [comment]
Would you be willing to merge in its current state if you did not think that it was a niche usage? Because the main place that I think this is useful is in serialized objects, such as in a database. It isn't *necessary* for that sort of thing, and is ugly to implement by hand, which is why it is rarely used. But it is a marked performance increase in a situatution like that, and I know that these situations are common.

**Comment by gappleto97** [comment]
Yeah, that makes total sense. I think a good approach would be to do something like a hybrid of these two patches. Make singleton be an argument to attr.s, but have it applied by this separate decorator. That way all the extra complexity is compartmentalized in a way that its easier to deal with.

That also has the advantage of being able to pair it with other types of data classes as well, so long as we can verify immutability somehow.

I think that the second branch avoided most of the performance worries that I had, while making the code a bit cleaner. Adding an argument to attr.s should be relatively short work, and I can do it tomorrow if you think that is reasonable.

**Comment by gappleto97** [comment]
So I added a bit to the alternate candidate (mainly namedtuple support), and I also added a commit to show what I had in mind for integration with `attr.s`. It is a 6 line change (excluding type stubs), just adding some defaults. If this is not desired, all you would need to do is revert commit f4382adc0461167548f89f5cbe6d189ff99843d2 before squashing/merging.

**Comment by hynek** [comment]
Closing, in the hope that #438 might be revived one day. :)
